src/libutil/hash.cc

```diff
@@ -180,63 +166,36 @@ Hash Hash::parseNonSRIUnprefixed(std::string_view s, HashAlgorithm algo)
 Hash::Hash(std::string_view rest, HashAlgorithm algo, bool isSRI)
     : Hash(algo)
 {
-    if (!isSRI && rest.size() == base16Len()) {
-
-        auto parseHexDigit = [&](char c) {
-            if (c >= '0' && c <= '9')
-                return c - '0';
-            if (c >= 'A' && c <= 'F')
-                return c - 'A' + 10;
-            if (c >= 'a' && c <= 'f')
-                return c - 'a' + 10;
-            throw BadHash("invalid base-16 hash '%s'", rest);
-        };
-
-        for (unsigned int i = 0; i < hashSize; i++) {
-            hash[i] = parseHexDigit(rest[i * 2]) << 4 | parseHexDigit(rest[i * 2 + 1]);
-        }
+    decltype(base16::decode) * decode;
+    std::string_view formatName;
+
+    if (isSRI) {
+        decode = base64::decode;
+        formatName = "SRI";
+    } else {
+        if (rest.size() == base16::encodedLength(hashSize)) {
+            decode = base16::decode;
+            formatName = "base16";
+        } else if (rest.size() == BaseNix32::encodedLength(hashSize)) {
+            decode = BaseNix32::decode;
+            formatName = "nix32";
+        } else if (rest.size() == base64::encodedLength(hashSize)) {
+            decode = base64::decode;
+            formatName = "base64";
+        } else
+            throw BadHash("hash '%s' has wrong length for hash algorithm '%s'", rest, printHashAlgo(this->algo));
     }
```

Comment:
Can this instead be an immediately invoked lambda:

```suggestion
    decltype(base16::decode) * decode;
    std::string_view formatName;
    
    auto [decode, formatName] = [&]() -> std::pair<decltype(base16::decode) *, std::string_view> {
      if (isSRI)
        return {base64::decode, "SRI"};
        
      if (rest.size() == base16::encodedLength(hashSize))
        return {base16::decode, "base16"};
        
      // other cases as well
    }();
```